pkg/roachprod/failureinjection/failures/failer.go

```diff
@@ -0,0 +1,202 @@
+	// i.e. we have recovered from a previous failure mode, or we never injected one.
+	readyForFailure
+	// activeFailure means there is an active FailureMode injected by this Failer.
+	// The Failer will not allow another FailureMode to be injected until the current
+	// one has been recovered. However, multiple different Failers can inject
+	// FailureModes concurrently.
+	activeFailure
+	// If something went wrong at any point during the FailureMode, we set the state to this.
+	// It is unclear what kind of state the cluster is in and steps may not necessarily be idempotent.
+	// Instead, this state indicates that the failure mode should be cleaned up for safety. Note this
+	// doesn't apply to WaitForFailureToPropagate or WaitForFailureToRecover, which shouldn't modify
+	// anything in the cluster, only monitor it.
+	invalid
+)
+
+func (s failureModeStage) String() string {
+	switch s {
+	case uninitialized:
+		return "uninitialized"
+	case readyForFailure:
+		return "readyForFailure"
+	case activeFailure:
+		return "activeFailure"
+	case invalid:
+		return "invalid"
+	default:
+		panic("unknown failure mode stage")
+	}
+}
+
+var invalidFailureStateErr = errors.New("failure mode is in invalid state, Cleanup() the failure mode and retry")
+
+// Failer is a wrapper over a FailureMode that enforces that we maintain
+// a consistent state in regards to a single failure mode usage. A Failer is
+// not thread-safe, and multiple Failers should be used for concurrent
+// failure injection instead.
+//
+// The following diagram illustrates the valid states and transitions:
+//
+//          ┌───────────────┐◄────────────────────────────┐
+//     ┌───►│ uninitialized │                             │
+//     │    └───────┬───────┴────────────────┐            │
+//     │            │                        │            │
+// Cleanup()     Setup()                     │        Cleanup()
+//     │            │                  Setup() Error      │
+//     │            │                        │            │
+//     │            │                        │            │
+//     │            │                        │            │
+//     │            ▼                        │            │
+//     │    ┌───────────────┐ Inject() Error │     ┌──────┴──────┐
+//     ├────┤readyForFailure├───────OR───────┼────►│   invalid   │◄───────┐
+//     │    └─┬─────────────┘Cleanup() Error │     └──────────┬──┘        │
+//     │      │          ▲                   │                │           │
+//     │      │          │                   │                │           │
+//     │      │          │            Recover() Error         │           │
+//     │   Inject()  Recover()               OR         Cleanup() Error───┘
+//     │      │          │            Cleanup() Error
+//     │      │          │                   │
+//     │      ▼          │                   │
+//     │    ┌───────────────┐                │
+//     └────┤ activeFailure ├────────────────┘
```

Comment:
Aren't we missing an edge between `activeFailure` and `invalid`? What if Recover() fails?

Reply:
Ah, maybe it's the dashed line which I misinterpreted as going into `uninitialized` :)